server/functions.lua

```diff
@@ -83,3 +83,143 @@ function ToggleDoorState(entity)
+---@param source number ID of the player
+---@param plate string The plate number of the vehicle.
+function public.giveKeysItem(source, plate)
+    local citizenid = getCitizenId(source)
+
+    if not citizenid then return end
+
+    local vehicleId = MySQL.single.await('SELECT id FROM player_vehicles WHERE citizenid = ? AND plate = ? LIMIT 1', { citizenid, plate })
+    exports.ox_inventory:AddItem(source, 'vehiclekeys', 1, { vehicleId = vehicleId, key = 'placeholder' })
+end
+
+---Takes the player the vehicle keys item.
+---@param source number ID of the player
+---@param plate string The plate number of the vehicle.
+function public.removeKeysItem(source, plate)
+    local citizenid = getCitizenId(source)
+
+    if not citizenid then return end
+
+    local vehicleId = MySQL.single.await('SELECT id FROM player_vehicles WHERE citizenid = ? AND plate = ? LIMIT 1', { citizenid, plate })
+    exports.ox_inventory:RemoveItem(source, 'vehiclekeys', 1, { vehicleId = vehicleId, key = 'placeholder' })
+end
+
+---Check if player has vehicle keys
+---@param source number ID of the player
+---@param vehicleId number unique id of the vehicle
+---@return boolean
+function public.hasKeysItem(source, vehicleId)
+    return exports.ox_inventory:GetItemCount(source, 'vehiclekeys', { vehicleId = vehicleId, key = 'placeholder' }, true) > 0
+end
+
+---Loads a players vehicles to the vehicleList
+---@param src integer
+function public.addPlayer(src)
+    local citizenid = getCitizenId(src)
+    if not citizenid then return end
+
+    local vehicles = MySQL.query.await('SELECT * FROM player_vehicles WHERE citizenid = ?', { citizenid })
+
+    local state = {}
+    local platesAssociations = {}
+
+    for i = 1, #vehicles do
+        platesAssociations[vehicles[i].plate] = true
+    end
+
+    local worldVehicles = GetAllVehicles()
+    for i = 1, #worldVehicles do
+        local vehiclePlate = qbx.getVehiclePlate(worldVehicles[i])
+        if platesAssociations[vehiclePlate] then
+            state[vehiclePlate] = true
+        end
+    end
+
+    Player(src).state:set('keysList', state, true)
+end
+
+---Removes a players vehicles from the vehicleList
+---@param src integer
+function public.removePlayer(src)
+    local citizenid = getCitizenId(src)
+    if not citizenid then return end
+
+    Player(src).state:set('keysList', nil, true)
+end
+
+--- Removing the vehicle keys from the user
+---@param source number ID of the player
+---@param plate string The plate number of the vehicle.
+function public.removeKeys(source, plate)
+    local citizenid = getCitizenId(source)
+
+    if not citizenid then return end
+
+    local keys = Player(source).state.keysList or {}
+
+    if not keys[plate] then return end
+    keys[plate] = nil
+
+    Player(source).state:set('keysList', keys, true)
+
+    exports.qbx_core:Notify(source, locale('notify.keys_removed'))
+
+    return true
+end
+
+function public.hasKeys(source, plate)
+    return Player(source).state.keysList[plate]
```

Comment:
Shouldn't this check for the existence of item based keys as well?

Reply:
As you mentioned earlier, it is better to put it in a separate branch. I gave these functions only as a sample

Reply:
Can we remove anything having to do with keys as items from this PR then?

Reply:
of course